http/src/main/java/com/hazelcast/jet/contrib/http/impl/HttpListenerSourceContext.java

```diff
@@ -0,0 +1,115 @@
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+
+import static io.undertow.Handlers.exceptionHandler;
+import static io.undertow.Handlers.path;
+import static io.undertow.util.Methods.CONNECT;
+import static io.undertow.util.Methods.DELETE;
+import static io.undertow.util.Methods.HEAD;
+import static io.undertow.util.Methods.OPTIONS;
+import static io.undertow.util.Methods.PATCH;
+import static io.undertow.util.Methods.TRACE;
+
+public class HttpListenerSourceContext<T> {
+
+    private final ILogger logger;
+    private final BlockingQueue<T> queue = new ArrayBlockingQueue<>(1024);
+    private final List<T> buffer = new ArrayList<>(1024);
+    private final Undertow undertow;
+    private final FunctionEx<byte[], T> mapToItemFn;
+
+    public HttpListenerSourceContext(
+            @Nonnull Processor.Context context,
+            int port,
+            @Nullable SupplierEx<SSLContext> sslContextFn,
+            @Nonnull FunctionEx<byte[], T> mapToItemFn
+    ) {
+        this.logger = context.logger();
+        this.mapToItemFn = mapToItemFn;
+        Address localAddress = context.jetInstance().getHazelcastInstance().getCluster().getLocalMember().getAddress();
+        String host = localAddress.getHost();
+        Undertow.Builder builder = Undertow.builder();
+        if (sslContextFn != null) {
+            builder.addHttpsListener(port, host, sslContextFn.get());
+            logger.info("Starting to listen HTTPS messages on https://" + host + ":" + port);
+        } else {
+            builder.addHttpListener(port, host);
+            logger.info("Starting to listen HTTP messages on http://" + host + ":" + port);
+        }
+        undertow = builder.setHandler(handler()).build();
+        undertow.start();
+    }
+
+    public void fillBuffer(SourceBuffer<T> sourceBuffer) {
+        queue.drainTo(buffer);
+        buffer.forEach(sourceBuffer::add);
+        buffer.clear();
+    }
+
+    public void close() {
+        undertow.stop();
+    }
+
+    private DisallowedMethodsHandler handler() {
+        return new DisallowedMethodsHandler(path().addExactPath("/", exceptionHandler(this::handleMainPath)
+                .addExceptionHandler(JsonProcessingException.class, this::handleJsonException)),
+                PATCH, HEAD, DELETE, CONNECT, OPTIONS, TRACE);
+    }
+
+    private void handleMainPath(HttpServerExchange exchange) {
+        exchange.getRequestReceiver().receiveFullBytes(this::consumeMessage, this::consumeException);
+    }
+
+    private void handleJsonException(HttpServerExchange exchange) {
+        logger.warning("Supplied payload is not a valid JSON: " +
+                exchange.getAttachment(ExceptionHandler.THROWABLE).getMessage());
+        exchange.setStatusCode(StatusCodes.BAD_REQUEST);
+    }
+
+    private void consumeMessage(HttpServerExchange exchange, byte[] message) {
+        queue.offer(mapToItemFn.apply(message));
```

Comment:
`offer` fails to insert the item and returns false in case the queue is full. 
Should use `put`?

Reply:
👍